<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <meta name="author" content="David Montaner" />
  <title>Practical session with BLAST: species</title>
  <style type="text/css">code{white-space: pre;}</style>
  <style>
  
  html {
      color:#000;
      background:#FFF;
  }
  
  body,div,dl,dt,dd,ul,ol,li,h1,h2,h3,h4,h5,h6,pre,code,form,fieldset,legend,input,textarea,p,blockquote,th,td {
      margin:0;
      padding:0;
  }
  
  table {
      border-spacing:0;
      font-size:inherit;
      font:100%;
      /*max-width:100%;*/
      background-color:transparent;
      border:1px solid #DDD;
      border-left:0;
      border-collapse:collapsed;
      -webkit-border-radius:4px;
      -moz-border-radius:4px;
      border-radius:4px;
  }
  
  fieldset,img {
      border:0;
  }
  
  address,caption,cite,code,dfn,em,strong,th,var {
      font-style:normal;
      font-weight:400;
  }
  
  ol,ul {
      list-style:none;
  }
  
  caption,th {
      text-align:left;
  }
  
  h1,h2,h3,h4,h5,h6 {
      font-size:100%;
      font-weight:400;
      color:#20639b;
      margin:1em 0;
  }
  
  q:before,q:after {
      content:'';
  }
  
  abbr,acronym {
      border:0;
      font-variant:normal;
      border-bottom:1px dotted #000;
      cursor:help;
  }
  
  sup {
      vertical-align:text-top;
  }
  
  sub {
      vertical-align:text-bottom;
  }
  
  input,textarea,select {
      font-family:inherit;
      font-weight:inherit;
      font-size:100%;
  }
  
  legend {
      color:#000;
  }
  
  h1,h2,h3,h4,h5,h6,strong {
      font-weight:700;
  }
  
  em {
      font-style:italic;
  }
  
  blockquote,ul,ol,dl {
      margin:1em;
      font-style:italic;
  }
  
  ol,ul,dl {
      margin-left:2em;
  }
  
  ol {
      list-style:decimal outside;
  }
  
  ul {
      list-style:disc outside;
  }
  
  dl dd {
      margin-left:1em;
  }
  
  th,td {
      line-height:1.125em;
      text-align:left;
      vertical-align:top;
      border-top:1px solid #DDD;
      border-left:1px solid #DDD;
      padding:.5em;
  }
  
  th {
      font-weight:700;
      text-align:center;
      color:#20639b;
  }
  
  table thead:first-child tr:first-child th,table tbody:first-child tr:first-child th,table tbody:first-child tr:first-child td {
      border-top:0;
  }
  
  table thead:first-child tr:first-child th:first-child,table tbody:first-child tr:first-child td:first-child {
      -webkit-border-radius:4px 0 0 0;
      -moz-border-radius:4px 0 0;
      border-radius:4px 0 0 0;
  }
  
  table thead:first-child tr:first-child th:last-child,table tbody:first-child tr:first-child td:last-child {
      -webkit-border-radius:0 4px 0 0;
      -moz-border-radius:0 4px 0 0;
      border-radius:0 4px 0 0;
  }
  
  table thead:last-child tr:last-child th:first-child,table tbody:last-child tr:last-child td:first-child {
      -webkit-border-radius:0 0 0 4px;
      -moz-border-radius:0 0 0 4px;
      border-radius:0 0 0 4px;
  }
  
  table thead:last-child tr:last-child th:last-child,table tbody:last-child tr:last-child td:last-child {
      -webkit-border-radius:0 0 4px 0;
      -moz-border-radius:0 0 4px;
      border-radius:0 0 4px 0;
  }
  
  table tbody tr:nth-child(odd) td,table tbody tr:nth-child(odd) th {
      background-color:#f9f9f9;
  }
  
  caption {
      margin-bottom:.5em;
      text-align:center;
  }
  
  p,fieldset,table,pre {
      margin-top:.5em;
      margin-bottom:.5em;
  }
  
  body {
      font:16px/1.4 "Helvetica Neue", Helvetica, "Open Sans", Arial, sans-serif;
      color:#333;
      /*max-width:960px;*/
      /*margin:0 auto;*/
      margin:1em;
  }
  
  pre,code,kbd,samp,tt {
      font-family:monospace;
      font-size:90%;
      line-height:100%;
  }
  
  h1 {
      font-size:200%;
      border-bottom:2px dotted #ddd;
  }
  
  a,a:visited,code {
      /*color:#f06024;*/
      color:#20639b;
  }
  
  a:link {text-decoration:none;}    /* unvisited link */
  a:visited {text-decoration:none;} /* visited link */
  a:hover {text-decoration:underline;}   /* mouse over link */
  a:active {text-decoration:underline;}  /* selected link */
  
  
  h2 {
      font-size:160%;
  }
  
  h3 {
      font-size:120%;
  }
  
  h4 {
      font-size:110%;
  }
  
  h5 {
      font-size:95%;
  }
  
  h6 {
      font-size:90%;
      color:#999;
  }
  
  dl {
      margin:0 0 1.125em;
  }
  
  dt {
      font-weight:700;
      line-height:1.125em;
  }
  
  dd {
      margin-left:1em;
      line-height:1.125em;
  }
  
  pre,blockquote {
      border:1px solid #ddd;
      background-color:#f1f1f1;
      -webkit-border-radius:6px;
      -moz-border-radius:6px;
      border-radius:6px;
      padding:10px;
  }
  
  blockquote {
      margin:1em 2.5em;
  }
  
  code {
      color:#20639b;
      background:#f1f1f1;
  }
  
  img {
      max-width:100%;
  }
  
  ol[type=1] {
      list-style-type:decimal;
  }
  
  ol[type=a] {
      list-style-type:lower-alpha;
  }
  
  ol[type=A] {
      list-style-type:upper-alpha;
  }
  
  ol[type=I] {
      list-style-type:upper-roman;
  }
  
  ol[type=i] {
      list-style-type:lower-roman;
  }
  
  
  
  h1.title {
          font-size: 200%;
          font-weight: bold;
          padding-top: 0.2em;
          /*padding-bottom: 0.2em;*/
          text-align: center;
          border: none;
  }
  
  h2.author {
          font-size: 100%;
          text-align: center;
          border: none;
  }
  
  h3.date {
          font-size: 75%;
          text-align: center;
          border: none;
          font-style:italic;
          font-weight: bold;
  }
  
  </style>
</head>
<body>
<div id="header">
<h1 class="title">Practical session with BLAST: species</h1>
<h2 class="author"><a href="http://www.dmontaner.com/">David Montaner</a></h2>
<h3 class="date"><em>(2014-07-11)</em></h3>
</div>
<div id="TOC">
<ul>
<li><a href="#preliminaries">Preliminaries</a><ul>
<li><a href="#software-used-in-this-practical">Software used in this practical</a></li>
</ul></li>
<li><a href="#exercise">Exercise</a><ul>
<li><a href="#explore-your-data">Explore your data</a></li>
<li><a href="#blast">BLAST</a></li>
<li><a href="#download-and-format">Download and Format</a></li>
<li><a href="#explore-the-tabular-results-using-r">Explore the tabular results using R</a></li>
</ul></li>
<li><a href="#homework">Homework</a></li>
</ul>
</div>
<!--
Clean the starting dataset
================================================================================
#    rm -r data/*
#    cp data0/f* data/
#    cd data
-->


<h1 id="preliminaries">Preliminaries</h1>
<p>The file <code>example.fa</code> in the <code>data</code> folder contains reads form the transcriptome of several <strong>different species</strong>. In this practical we will use BLAST to find out which species where present in our sample and to estimate the percentage of reads that belong to each of them.</p>
<p>First we will carry out a BLAST analysis using the <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi" title="BLAST web application at the NCBI">NCBI web application</a>. Then we will use the local application <code>blast_formatter</code> to download the data from the web and to shape them in the most appropriated format. Finally we will use <a href="http://r-project.org/" title="R project home page">R</a> to estimate the percentages of each species.</p>
<h2 id="software-used-in-this-practical">Software used in this practical</h2>
<ul>
<li>Linux Shell</li>
<li>blast_formatter</li>
<li><a href="http://r-project.org/" title="R project home page">R</a></li>
</ul>
<h1 id="exercise">Exercise</h1>
<h2 id="explore-your-data">Explore your data</h2>
<p>Go to the data directory and have a look of the first lines of the file</p>
<pre><code>cd data
head example.fa </code></pre>
<p>It is a standard fasta file. You can find out the number of reads or sequences in the file using the option <code>-c</code> in the grep command</p>
<pre><code>grep -c &quot;^&gt;&quot; example.fa </code></pre>
<p>or using the <code>wc</code> command after the <code>grep</code></p>
<pre><code>grep &quot;^&gt;&quot; example.fa | wc -l </code></pre>
<p>You could dump all the sequence identifier of the fasta file in a separated file doing</p>
<pre><code>grep &quot;^&gt;&quot; example.fa &gt; myids.txt
sed -i &#39;s/&gt;//&#39; myids.txt</code></pre>
<ul>
<li>What does it mean the <code>^</code> symbol in the grep expression?</li>
<li>Is it necessary?</li>
<li>What does the command <code>seed</code> do?</li>
</ul>
<h2 id="blast">BLAST</h2>
<p>Go to the <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi" title="BLAST web application at the NCBI">NCBI web application</a> and upload the <code>example.fa</code> to run a <strong>blastn</strong></p>
<ul>
<li>Why are we running <em>blastn</em> and not <em>blastp</em> or <em>blastx</em>?</li>
<li>Which is the data base we are blasting against?</li>
</ul>
<p>Running the blast with his amount of data usually takes around a minute.</p>
<p>Explore the HTML results when the blast is finished.</p>
<ul>
<li>How good do they seem the alignments?</li>
<li>What does it mean the <strong>RID</strong> code? <!-- BLAST search Request ID--></li>
<li>How long are your results going to be kept in at the NCBI web site?</li>
</ul>
<p>Keep thee <strong>RID</strong> for later.</p>
<p>Download the results in the different formats you can and explore them.</p>
<ul>
<li>In which of the formats do we have information about the species to which our reads belong to?</li>
<li>Compare among them the sizes of the files you have downloaded.</li>
</ul>
<h2 id="download-and-format">Download and Format</h2>
<p>From the web page you can save just a few of the output formats available in BLAST. The command line function <code>blast_formatter</code> let us download many other, and even customize some of them.</p>
<p>If you see the help of the command</p>
<pre><code>blast_formatter -help </code></pre>
<p>You will find</p>
<pre><code>-outfmt &lt;String&gt;
  alignment view options:
    0 = pairwise,
    1 = query-anchored showing identities,
    2 = query-anchored no identities,
    3 = flat query-anchored, show identities,
    4 = flat query-anchored, no identities,
    5 = XML Blast output,
    6 = tabular, 
    7 = tabular with comment lines,
    8 = Text ASN.1,
    9 = Binary ASN.1,
   10 = Comma-separated values,
   11 = BLAST archive format (ASN.1) </code></pre>
<p>We can use <code>blast_formatter</code> tool to download a <strong>tab</strong> separated file with the BLAST results doing</p>
<pre><code>blast_formatter -rid M4DNEDCY01R -outfmt 6 -out my_tab_blast.txt</code></pre>
<p>This tabular format is easy to handle and takes small amount of memory compared with some of the other files, but the <strong>species</strong> of the nucleotide sequences in the data base are not there.</p>
<p>To get this extra information we can customize the format of the <code>blast_formatter</code> output. See the help of the command.</p>
<pre><code>blast_formatter -rid M4DNEDCY01R -outfmt &quot;6 qseqid sgi sacc saccver staxids&quot; -out my_tab_blast_and_species.txt</code></pre>
<!-- 
Estos campos (del final) devuelven NAs ???

    blast_formatter -rid M4DNEDCY01R -outfmt "6 qseqid sgi sacc saccver score staxids" -out my_tab_blast_and_species.txt

    blast_formatter -rid M4DNEDCY01R -outfmt "6 qseqid sgi staxids sscinames scomnames sblastnames sskingdoms" -out my_tab_blast_and_species.txt

--> 




<h2 id="explore-the-tabular-results-using-r">Explore the tabular results using R</h2>
<p>Use R to read the files <code>my_tab_blast.txt</code> and <code>my_tab_blast_and_species.txt</code>.</p>
<p>Merge them and find the <strong>best hit</strong> (the one that maximizes the <em>bit score</em>, provided that the <em>e-value</em> surpasses certain quality threshold) for each read.</p>
<p>See the file <code>explore_blast_tables.r</code> for details. The output of the script is <a href="explore_blast_tables.html">here</a>.</p>
<h1 id="homework">Homework</h1>
<p>Find your dataset in the folder <a href="datos">datos</a> and find which are the <strong>tree</strong> species more abundant in it.</p>
<p>Which are the concentrations of each of them in percentage?</p>
</body>
</html>
